### PR TITLE
UserAvatarLink widget

### PR DESCRIPTION
**Pull request by Ngoakor12**
Related issues: [[1. Create a new widget called UserAvatarLink](https://www.notion.so/1-Create-a-new-widget-called-UserAvatarLink-7588bf3fc6ba46548c9e18307d45373f)]

## Description:

What are you up to? Fill us in :)

## Screenshots/videos
For umuzi emails
image
Other emails
image
image

## I solemnly swear that:

- [x] My code follows the style guidelines of this project
- [x] I have merged the develop branch into my branch and fixed any merge conflicts
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: frontend/src/components/widgets/UserAvatarLink.jsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  },
+  container: {
+    display: "flex",
+    alignItems: "center",
+    gap: 8,
+  },
+});
+
+function getAvatarTextInitials(email) {
+  if (email.endsWith("umuzi.org")) {
```
Comment (FaithMo) [inline frontend/src/components/widgets/UserAvatarLink.jsx:22]: there are other learners who are in the bridge course or in BootCamp, and their emails are not umuzis.

File: frontend/src/components/widgets/UserAvatarLink.jsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    const [name, surname] = email.split(".");
+    return (name[0] + surname[0]).toUpperCase();
+  }
+  const firstTwoLetters = email.match(/^[A-Za-z]{2}/);
```
Comment (FaithMo) [inline frontend/src/components/widgets/UserAvatarLink.jsx]: ...and some people who use their personal emails use funky names

**Review by FaithMo — CHANGES_REQUESTED** [review]

File: frontend/src/components/widgets/UserAvatarLink.jsx (hunk 2 above)
Comment (Ngoakor12) [reply frontend/src/components/widgets/UserAvatarLink.jsx]: It's supposed to be working for both umuzi emails and other funky emails people may have.

I can add tests for this function for verification though and test it with different emails.

**Comment by Ngoakor12** [comment]
@FaithMo I added tests to verify that it would work for other emails and they all pass. I ran the tests in a separate project with tests since there are failing tests for this project(develop branch).

**Review by sheenarbw — CHANGES_REQUESTED** [review]
Can you please add a screenshot to visual PRs? It makes review much nicer.

**Comment by Ngoakor12** [comment]
screenshots added

File: frontend/src/components/widgets/UserAvatarLink.test.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { getAvatarTextInitials } from "./UserAvatarLink";
+
+it("getAvatarTextInitials function returns correct initials from an umuzi.org email", () => {
+  expect(getAvatarTextInitials("ngoako.ramokgopa@umuzi.org")).toBe("NR");
+  expect(getAvatarTextInitials("faith.mofokeng@umuzi.org")).toBe("FM");
+});
+
+it("getAvatarTextInitials function returns correct initials from other emails", () => {
+  expect(getAvatarTextInitials("projectlkh63@gmail.com")).toBe("PR");
+  expect(getAvatarTextInitials("c.nyatondo@gmail.com")).toBe("NY");
```
Comment (sheenarbw) [inline frontend/src/components/widgets/UserAvatarLink.test.js]: Why not:

```
  expect(getAvatarTextInitials("c.nyatondo@gmail.com")).toBe("CN");
  expect(getAvatarTextInitials("n0606160@gmail.com")).toBe("N0");
  expect(getAvatarTextInitials("08606160@gmail.com")).toBe("08");
```
This seems more descriptive and useful.

**Review by sheenarbw — CHANGES_REQUESTED** [review]
Mostly good. Some of your choices about what letters to show are a bit odd though

File: frontend/src/components/widgets/UserAvatarLink.jsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    display: "flex",
+    alignItems: "center",
+    gap: 8,
+  },
+});
+
+function getAvatarTextInitials(email) {
+  // For Umuzi emails - take the first letters of the name and surname
+  if (email.endsWith("umuzi.org")) {
+    const [firstname, lastname] = email.split(".");
```
Comment (FaithMo) [inline frontend/src/components/widgets/UserAvatarLink.jsx]: I don't know if this is an issue, but if we were to ever want to use the last name for something else, this variable `lastname` would have to be broken down further. for example, for `ngoako.ramokgopa@umuzi.org` the lastname is ramokgopa@umuzi. but even if not, I think the variable name has to tell the truth about its value.

**Review by FaithMo — APPROVED** [review]

File: frontend/src/components/widgets/UserAvatarLink.jsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    textDecoration: "none",
+  },
+  container: {
+    display: "flex",
+    alignItems: "center",
+    gap: 8,
+  },
+});
+
+export function getAvatarTextInitials(email) {
```
Comment (Vuyisanani) [inline frontend/src/components/widgets/UserAvatarLink.jsx]: Great work,  just a bit concerned about the choice of names (variables  and functions)

Shouldn't this be rather getAvatarInitials?

File: frontend/src/components/widgets/UserAvatarLink.jsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  const customPartOfAnEmail = email.substr(0, email.indexOf("@"));
```
Comment (Vuyisanani) [inline frontend/src/components/widgets/UserAvatarLink.jsx]: customPartOfAnEmail ?? I shook my head around for some time, but what I see is you getting the prefix of an email

If my email is vuyisanani.meteni@umuzi.org,  you only getting/storing vuyisanani.meteni in your variable, not entirely sure if your variable more descriptive for that

File: frontend/src/components/widgets/UserAvatarLink.jsx [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  // For Umuzi emails - take the first letters of the firstname and lastname
+  if (email.endsWith("umuzi.org")) {
+    const [firstname, lastname] = customPartOfAnEmail.split(".");
+    return (firstname[0] + lastname[0]).toUpperCase();
+  }
+  // Other emails - take first two letters/digits
+  let firstTwoLetters = "";
+  const alphanumerals = /[a-zA-Z0-9]/;
```
Comment (Vuyisanani) [inline frontend/src/components/widgets/UserAvatarLink.jsx]: alphanuerals regex is a const and won't be changed if you were to reuse it somewhere on top of your code, you would have to redeclare, would you perhaps rather globally initialize this variable?

**Review by Vuyisanani — CHANGES_REQUESTED** [review]

File: frontend/src/components/widgets/UserAvatarLink.jsx (hunk 6 above)
Comment (Vuyisanani) [reply frontend/src/components/widgets/UserAvatarLink.jsx]: perhaps use `emailUsername`

**Review by Vuyisanani — CHANGES_REQUESTED** [review]

File: frontend/src/components/widgets/UserAvatarLink.jsx (hunk 6 above)
Comment (Ngoakor12) [reply frontend/src/components/widgets/UserAvatarLink.jsx]: emailUsername is perfect

**Review by FaithMo — APPROVED** [review]

**Review by Vuyisanani — APPROVED** [review]

**Review by sheenarbw — APPROVED** [review]

**Pull request merged**